add spark version

juliangruber opened the pull request:
Add spark client version, so that we can create retrieval status graphs where we filter out versions that are known to be broken.

Related: https://github.com/filecoin-station/spark/pull/7

bajtos commented:
I love the idea of detecting broken SPARK versions that are not able to perform retrievals correctly 👍🏻 

I have a few ideas to consider:

(1)
>  create retrieval status graphs where we filter out versions that are known to be broken.

If all we want is filtering out reported data, I think adding the version to the retrieval stats table is much simpler. Adding one more field is, IMO, easier than adding logic to handle the request body when creating a new job.

(2)
SPARK version is often not enough to decide whether a retrieval failed because of a known bug. Recently, the problem was in the Zinnia runtime, and we fixed it by producing a new version of Zinnia, not SPARK. Maybe we should report both SPARK and Zinnia versions?

Since this change requires Zinnia to provide its version to modules, it may be better to leave it out of the initial PR. On the other hand, we can add the zinnia_version field to our API and DB now if we make it optional. That way we don't need to touch this repo later.

(3) 
On second thought, I like your proposal to report the version when creating a new retrieval job. This enables us to reject such requests coming from outdated SPARK versions because what's the point of creating a job when we know the retrieval will fail because of a bug in our code? When we reject a request to create a new job, the response can indicate that the SPARK client is outdated. In SPARK, we can convert this error into a Station activity asking the user to upgrade their Station.

I don't want to over-engineer this, but at the same time, I feel taking another look and doing one more iteration could greatly improve the solution at very little cost.

Let me know which of my ideas resonate with you. There is nothing wrong with your current proposal, we can defer my suggestions to future iterations.

bajtos commented on test/test.js:
```diff
@@ -360,10 +383,11 @@ describe('Routes', () => {
+      const createRequest = await fetch(`${spark}/retrievals`, {
+        method: 'POST',
+        headers: { 'Content-Type': 'application/json' },
+        body: JSON.stringify({ sparkVersion: '1.2.3' })
```
Do you have a test to verify that `sparkVersion` is not required?

It would be nicer if you could keep this test with no changes to prove that you are not introducing any regressions and then add a new test focused on handling the new field `sparkVersion`.

The new test can be super simple, as we don't care about any request/response fields  except `sparkVersion`.

Also, do you have a test to verify that `sparkVersion` is persisted in the database? The best place for this is IMO in the new test I described above.

I know that we have different views on how much tests to write. Please pick the parts of my comment that you find useful and feel free to ignore the rest.

juliangruber replied:
> Do you have a test to verify that sparkVersion is not required?

Yes, this is the `POST /retrieval > creates a retrieval` test.

> It would be nicer if you could keep this test with no changes to prove that you are not introducing any regressions and then add a new test focused on handling the new field sparkVersion.

This is covered by the `POST /retrieval` tests already, here we're not testing whether we can create proper retrievals, but whether we can read them correctly.

> Also, do you have a test to verify that sparkVersion is persisted in the database? The best place for this is IMO in the new test I described above.

Yes, this is the `POST /retrievals > handles versions` test.

bajtos reviewed (commented):
I reviewed the code and have a few suggestions to consider.

juliangruber commented on test/test.js:
```diff
@@ -26,7 +25,13 @@ describe('Routes', () => {
+      client,
+      logger: {
+        info () {},
+        error (...args) { console.error(...args) }
```
I found an issue with `noop-logger`: Ignoring all logs was bad, because if something errored badly you wouldn't see a needed log line in the test output. This changes it so that only expected log lines are hidden

juliangruber commented:
> If all we want is filtering out reported data, I think adding the version to the retrieval stats table is much simpler. Adding one more field is, IMO, easier than adding logic to handle the request body when creating a new job.

Right, we were doing the same for `walletAddress` already. Updated!

> SPARK version is often not enough to decide whether a retrieval failed because of a known bug. Recently, the problem was in the Zinnia runtime, and we fixed it by producing a new version of Zinnia, not SPARK. Maybe we should report both SPARK and Zinnia versions?
> 
> Since this change requires Zinnia to provide its version to modules, it may be better to leave it out of the initial PR. On the other hand, we can add the zinnia_version field to our API and DB now if we make it optional. That way we don't need to touch this repo later.

Great idea! I fully agree, and added the optional column `zinnia_version`

> On second thought, I like your proposal to report the version when creating a new retrieval job. This enables us to reject such requests coming from outdated SPARK versions because what's the point of creating a job when we know the retrieval will fail because of a bug in our code? When we reject a request to create a new job, the response can indicate that the SPARK client is outdated. In SPARK, we can convert this error into a Station activity asking the user to upgrade their Station.

I should have finished reading your comment before jumping into the implementation :D Hmm this is a good point and a tough call. I think you're right. I've migrated them back, I think this will be a useful feature to have.

juliangruber commented on index.js:
```diff
@@ -3,10 +3,22 @@ import { migrate } from './lib/migrate.js'
+  }
+  if (type === 'date') {
+    const date = new Date(obj[key])
+    assert(!isNaN(date.getTime()), 400, `Invalid .${key}`)
```
58ed6a6

juliangruber commented on index.js:
```diff
@@ -3,10 +3,22 @@ import { migrate } from './lib/migrate.js'
+    const date = new Date(obj[key])
+    assert(!isNaN(date.getTime()), 400, `Invalid .${key}`)
+  } else {
+    assert.strictEqual(typeof obj[key], type, 400, `Invalid .${key}`)
```
58ed6a6

juliangruber commented on index.js:
```diff
@@ -3,10 +3,22 @@ import { migrate } from './lib/migrate.js'
+  if (!required && (!Object.keys(obj).includes(key) || obj[key] === null)) {
+    return
+  }
+  if (type === 'date') {
```
What about this? 79e2b21

bajtos reviewed (approved):
👌🏻
